pkg/controller/replicaset/replica_set_test.go

```diff
@@ -1982,6 +2005,99 @@ func TestGetPodsToDelete(t *testing.T) {
+	return newPod
+}
+
+func TestComparingSameNodeSpreadOnDeletion(t *testing.T) {
+	labelMap := map[string]string{"name": "foo"}
+	rs := newReplicaSet(6, labelMap)
+	timeADayAgo := time.Now().Add(-time.Hour * 24)
+
+	// scheduled, running, ready pods on fake-node-1
+	podAOnNode1 := newPod("pod-a-on-node-1", rs, v1.PodRunning, &metav1.Time{Time: timeADayAgo.Add(-time.Hour)}, true)
+	podBOnNode1 := newPod("pod-b-on-node-1", rs, v1.PodRunning, &metav1.Time{Time: timeADayAgo.Add(time.Hour)}, true)
+	podAOnNode1.Spec.NodeName, podBOnNode1.Spec.NodeName = "fake-node-1", "fake-node-1"
+
+	// scheduled, running, ready pods on fake-node-2
+	podAOnNode2 := newPod("pod-a-on-node-2", rs, v1.PodRunning, &metav1.Time{Time: timeADayAgo.Add(-time.Hour * 6)}, true)
+	podBOnNode2 := newPod("pod-b-on-node-2", rs, v1.PodRunning, &metav1.Time{Time: timeADayAgo.Add(time.Hour * 2)}, true)
+	podCOnNode2 := newPod("pod-c-on-node-2", rs, v1.PodRunning, &metav1.Time{Time: timeADayAgo.Add(-time.Hour * 2)}, true)
+	podAOnNode2.Spec.NodeName, podBOnNode2.Spec.NodeName, podCOnNode2.Spec.NodeName = "fake-node-2", "fake-node-2", "fake-node-2"
+
+	// scheduled, running, ready pods on fake-node-3
+	podAOnNode3 := newPod("pod-a-on-node-3", rs, v1.PodRunning, &metav1.Time{Time: timeADayAgo.Add(time.Hour * 6)}, true)
+	podAOnNode3.Spec.NodeName = "fake-node-3"
+
+	// current state: node1(a,b), node2(a,b,c), node3(a)
+	// when this rs scales to 3, we expect to get one pod left on each node, which should be a balanced state and best for fault tolerance.
+
+	tests := []struct {
+		name           string
+		rankFunc       func(podsToRank, relatedPods []*v1.Pod) controller.ActivePodsWithRanks
+		interval       time.Duration
+		expectedSpread bool
+	}{
+		{
+			name: "sort with the old getPodsRankedByRelatedPodsOnSameNode",
+			rankFunc: func(podsToRank, relatedPods []*v1.Pod) controller.ActivePodsWithRanks {
+				podsOnNode := make(map[string]int)
+				for _, pod := range relatedPods {
+					if controller.IsPodActive(pod) {
+						podsOnNode[pod.Spec.NodeName]++
+					}
+				}
+				ranks := make([]int, len(podsToRank))
+				for i, pod := range podsToRank {
+					ranks[i] = podsOnNode[pod.Spec.NodeName]
+				}
+				return controller.ActivePodsWithRanks{Pods: podsToRank, Rank: ranks, Now: metav1.Now()}
+			},
+			expectedSpread: false,
+		},
+		{
+			name: "sort without extra ranks (just rely on the random timestamp sort)",
+			rankFunc: func(podsToRank, relatedPods []*v1.Pod) controller.ActivePodsWithRanks {
+				return controller.ActivePodsWithRanks{Pods: podsToRank, Rank: make([]int, len(podsToRank)), Now: metav1.Now()}
+			},
+			interval:       time.Millisecond * 5,
+			expectedSpread: false,
+		},
+		{
+			name:           "sort with the new getPodsRankedByRelatedPodsOnSameNode",
+			rankFunc:       getPodsRankedByRelatedPodsOnSameNode,
+			expectedSpread: true,
+		},
+	}
+
+	for _, test := range tests {
+		// In consequence of random sort, each test will run 100 times.
```

Comment:
The "random" sort is not really random. We use the UUID to compare them, since that's a random value. Since you are using the same pods in all tests, you will get the same result.

To make it random, you would have to create the pods in each iteration.